[TIMOB-18565] Add support for Android Libraries

janvennemann opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-18565

**Description:**
This allows Android Libraries in the form of Android Archive (AAR)
files to be dropped in as module dependencies.

janvennemann committed:
[TIMOB-18565] Add support for Android Libraries

This allows Android Libraries in the form of Android Archive (AAR)
files to be dropped in as module dependencies.

AppWerft commented:
nice. But what is with this problem: the third party aar uses more then one R.class. Is it solved with this PR?

janvennemann commented:
@AppWerft, do you happen to have an example .aar to test against? Currently only the main package defined in AndroidManifest.xml from inside the .aar will be passed as an extra package to AAPT.

AppWerft commented:
The aar is not open source. Inside are res for library, for v4 and for v7. Currently we have merged all stuff in one res. But in classes are references to 3 *.*.R. I have xinged you.

janvennemann committed:
[TIMOB-18565] Add AAR transform library

janvennemann committed:
[TIMOB-18565] Show warning on duplicate resource files

garymathews commented on android/cli/hooks/aar-transform.js:
```diff
@@ -0,0 +1,45 @@
+	cli.on('build.module.pre.compile', {
+		post: function (builder, callback) {
+			var foundAndroidArchives = [];
+			var aarOutputPath = path.join(builder.buildDir, 'intermediates/exploded-aar');
```
```Javascript
var aarOutputPath = path.join(builder.buildDir, 'intermediates', 'exploded-aar');
var assetsDestinationPath = path.join(builder.buildDir, 'intermediates', 'assets');
var libraryDestinationPath = path.join(builder.buildDir, 'intermediates', 'lib');
var sharedLibraryDestinationPath = path.join(builder.buildGenJniLocalDir, 'jni');
```

hansemannn committed:
Merge branch 'master' into TIMOB-18565

janvennemann committed:
[TIMOB-18565] Update dependency appc-aar-tools

janvennemann committed:
[TIMOB-18565] Improve .aar handling

The hook will now perform additional checks on the .aar files like
checking for a unique package names, skipping duplicate files based
on SHA-256 hash and cache the transform results to improve performance
on incremental builds.

janvennemann committed:
[TIMOB-18565] Update appc-aar-tools to 1.1.1

janvennemann committed:
[TIMOB-18565] Adopt Android Library handling from Gradle

This will change the handling of Android Libraries to be more like
Gradle does it. Instead of pre-merging resources, copying .jar and .so
libraries from every Android Archive, and package that into the final
module build, we only extract the Android Archive and pass those files
to the appropriate module build steps. The Android Archive will then be
packaged as is into the module and the app builder can process the
archive again.

janvennemann committed:
[TIMOB-18565] Manually create module .jar

janvennemann committed:
[TIMOB-18565] Remove module merging restrictions

The previous check wether a file is a resource file or not was based
on a very restricted set of filenames. This is not suitable anymore so
we now check for the existence of a resource tag as the root node in a
.xml file and treat everything as a resource file that does.

janvennemann commented:
@garymathews, updated the PR with my latest changes to the module build. The app build requires some more work and will follow tomorrow.

janvennemann committed:
[TIMOB-18565] Process resources from Android Libraries

janvennemann committed:
[TIMOB-18565] Add bundled libraries from android archive files to build

garymathews reviewed (approved):
CR: **PASS** after minor change
FT: **PASS**

###### TEST CASE
- Compiled a number of modules with AAR and JAR libs using:
```
 appc run -p android --build-only -l trace
```
and verified their output and usage.

AWESOME!! 🎆

garymathews commented on android/cli/hooks/aar-transform.js:
```diff
@@ -0,0 +1,446 @@
+				var fileReadStream = fs.createReadStream(aarPathAndFilename);
+				fileReadStream.on('readable', function() {
+					var data = fileReadStream.read();
+					if (data)
```
minor, but we should use brackets here

janvennemann committed:
[TIMOB-18565] Automatically crunch PNGs using AAPT

We do not manually crunch the PNG beforehand like in Gradle, so AAPT has to do it.

janvennemann committed:
[TIMOB-18565] Add missing curly brackets around if-block

janvennemann commented:
@garymathews, addressed the wrong `--no-crunch` option and missing brackets.

garymathews reviewed (approved):
CR: **PASS**
FT: **PASS**

👍

janvennemann committed:
[TIMOB-18565] Update appc-aar-tools to 1.1.2

hansemannn commented:
@janvennemann Please resolve the merge conflicts

janvennemann committed:
Merge remote-tracking branch 'upstream/master' into TIMOB-18565

williamrijksen commented:
@janvennemann Cool stuff!

hansemannn commented:
@janvennemann Please resolve the merge conflicts (😁). Should be the last one, but we moved the `titanium` library into an own `node-titanium-sdk` one, so this causing issues when requiring. Thanks!

janvennemann committed:
Merge remote-tracking branch 'upstream/master' into TIMOB-18565

janvennemann committed:
[TIMOB-18565] Use node-titanium-sdk instead of titanium-sdk

The titanium-sdk dependency was moved to its own repo and renamed to node-titanium-sdk

janvennemann committed:
[TIMOB-18565] Update appc-aar-tools to 1.1.3

hansemannn commented on android/cli/commands/_build.js:
```diff
@@ -35,8 +35,8 @@ var ADB = require('node-titanium-sdk/lib/adb'),
-	ti = require('titanium-sdk'),
-	tiappxml = require('titanium-sdk/lib/tiappxml'),
+	ti = require('node-titanium-sdk'),
+	tiappxml = require('node-titanium-sdk/lib/tiappxml'),
```
Can we please scan **all** usages for this require-issue? This is the 3rd commit where we need to manually adjust the path after initially changing it. @cb1kenobi ?

cb1kenobi replied:
Yeah, I think someone should just grep for `titanium-sdk` and replace with `node-titanium-sdk`. There shouldn't be 100's of instances.

hansemannn replied:
@janvennemann Can you do that? I'll be out until Wednesday and we should really get that in before we forget it. Maybe in an own PR which we can merge before this one gets queued. Thanks guys!

garymathews committed:
Merge branch 'master' into TIMOB-18565

Commit:
Merge branch 'master' into TIMOB-18565

Commit:
Merge branch 'master' into TIMOB-18565

Commit:
Merge branch 'master' into TIMOB-18565

lokeshchdhry commented:
FR done for backport PR.
Merging.
